core/utils/histogram.h

```diff
@@ -1,55 +1,145 @@
 #ifndef BESS_UTILS_HISTOGRAM_H_
 #define BESS_UTILS_HISTOGRAM_H_
 
-#include "time.h"
+#include <cmath>
+#include <cstring>
+#include <sstream>
+#include <string>
+#include <vector>
 
-#define HISTO_TIMEUNIT_MULT (1000lu * 1000 * 1000)  // Nano seconds
-#define HISTO_TIME (100lu)       // We measure in 100 ns units
-#define HISTO_BUCKETS (1000000)  // Buckets
-#define HISTO_HARD_TIMEOUT (HISTO_BUCKETS)
+static const std::vector<double> quartiles = {0.25f, 0.5f, 0.75f, 1.0f};
 
-#define HISTO_TIME_TO_SEC(t) ((t) / (HISTO_TIMEUNIT_MULT / HISTO_TIME))
-#define PRINT_THRES (1)
-#define HISTO_BUCKET_VAL(ptr) ((*(ptr)))
-#define HISTO_BUCKET_INC(ptr) ((*(ptr))++)
+// A general purpose histogram. A bin b_i is labeled by its upper edge
+// T must be an arithmetic type
+template <typename T>
+class Histogram {
+ public:
+  Histogram(size_t num_buckets, T threshold, T bucket_width)
```

Comment:
Definitely. There's actually quite a bit of documentation i want to add. The interface exposed is not entirely obvious